api/internal/handler/route/route_test.go

```diff
@@ -14,1743 +14,1421 @@
+
+			h := Handler{routeStore: mStore, scriptStore: sStore}
+			ctx := droplet.NewContext()
+			ctx.SetInput(tc.giveInput)
+
+			ret, err := h.List(ctx)
+			assert.Equal(t, tc.called, getCalled)
+			assert.Equal(t, tc.wantRet, ret)
+			assert.Equal(t, tc.wantErr, err)
+		})
+	}
+}
 
-	//get route -- deleted, not found
-	getInput := &GetInput{}
-	reqBody = `{"id": "1"}`
-	err = json.Unmarshal([]byte(reqBody), getInput)
-	assert.Nil(t, err)
-	ctx.SetInput(getInput)
-	ret, err = handler.Get(ctx)
-	assert.EqualError(t, err, "data not found")
-	assert.Equal(t, http.StatusNotFound, ret.(*data.SpecCodeResponse).StatusCode)
+func TestRoute_Create(t *testing.T) {
+	scriptMap := make(map[string]interface{})
 
-	//delete test data
-	reqBody = `{"ids": "not-exists"}`
-	err = json.Unmarshal([]byte(reqBody), inputDel)
+	err := json.Unmarshal([]byte(DagScript), &scriptMap)
 	assert.Nil(t, err)
-	ctx.SetInput(inputDel)
-	ret, err = handler.BatchDelete(ctx)
-	assert.NotNil(t, err)
-	assert.Equal(t, http.StatusNotFound, ret.(*data.SpecCodeResponse).StatusCode)
 
-	//create route with not exist upstream id
-	route4 := &entity.Route{}
-	reqBody = `{
-	  "id": "2222",
-	  "name": "r222",
-	  "uris": ["/aa", "/bb"],
-	  "upstream_id": "not-exists"
-	}`
-	err = json.Unmarshal([]byte(reqBody), route4)
+	luaCode, err := generateLuaCode(scriptMap)
 	assert.Nil(t, err)
-	ctx.SetInput(route4)
-	ret, err = handler.Create(ctx)
-	assert.NotNil(t, err)
-	assert.Equal(t, http.StatusBadRequest, ret.(*data.SpecCodeResponse).StatusCode)
 
-	//type:chash, hash_on: vars, wrong key
-	route5 := &entity.Route{}
-	reqBody = `{
-	  "id": "1",
-	  "methods": ["GET"],
-	  "upstream": {
-	      "nodes": {
-	          "127.0.0.1:8080": 1
-	      },
-	      "type": "chash",
-	      "hash_on":"vars",
-	      "key": "not_support"
-	  },
-	  "desc": "new route",
-	  "uri": "/index.html"
-	}`
-	err = json.Unmarshal([]byte(reqBody), route5)
-	assert.Nil(t, err)
-	ctx.SetInput(route5)
-	ret, err = handler.Create(ctx)
-	assert.NotNil(t, err)
-	assert.Equal(t, http.StatusBadRequest, ret.(*data.SpecCodeResponse).StatusCode)
+	tests := []testCase{
+		{
+			caseDesc: "create route success",
+			giveInput: &entity.Route{
+				BaseInfo: entity.BaseInfo{
+					ID:         "s1",
+					CreateTime: 1609746531,
+				},
+				Name:       "s1",
+				Desc:       "test_route",
+				UpstreamID: "u1",
+				ServiceID:  "s1",
+				Script:     "",
+				Labels: map[string]string{
+					"version": "v1",
+				},
+			},
+			mockInput: &entity.Route{
+				BaseInfo: entity.BaseInfo{
+					ID:         "s1",
+					CreateTime: 1609746531,
+				},
+				Name:       "s1",
+				Desc:       "test_route",
+				UpstreamID: "u1",
+				ServiceID:  "s1",
+				ScriptID:   "s1",
+				Script:     "",
+				Labels: map[string]string{
+					"version": "v1",
+				},
+			},
+			serviceRet:  "service",
+			serviceErr:  nil,
+			upstreamRet: "upstream",
+			upstreamErr: nil,
+			scriptRet:   "script",
+			scriptErr:   nil,
+			wantRet:     nil,
```

Comment:
I have a question: 
why the wantret nil returned?
This is not consistent with the interface expectation. 
Normally, it should return data.

Reply:
+1. We shall keep the behavior of the mock object same as the real one.

Reply:
Good capture ! Let me fix this, and check other testcase.

Reply:
Fixed. Please review again. @Jaycean @imjoey